lib/services/local/inventory.go

```diff
@@ -0,0 +1,149 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package local
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	apidefaults "github.com/gravitational/teleport/api/defaults"
+	"github.com/gravitational/teleport/api/internalutils/stream"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// GetInstances iterates all teleport instances.
+func (s *PresenceService) GetInstances(ctx context.Context, req types.InstanceFilter) stream.Stream[types.Instance] {
+	const pageSize = 10_000
+	if req.ServerID != "" {
+		instance, _, err := s.GetRawInstance(ctx, req.ServerID)
+		if err != nil {
+			if trace.IsNotFound(err) {
+				return stream.Empty[types.Instance]()
+			}
+			return stream.Fail[types.Instance](trace.Wrap(err))
+		}
+		if !req.Match(instance) {
+			return stream.Empty[types.Instance]()
+		}
+		return stream.Once(instance)
+	}
+
+	startKey := backend.Key(instancePrefix, "")
+	endKey := backend.RangeEnd(startKey)
+	items := backend.StreamRange(ctx, s, startKey, endKey, pageSize)
+	return stream.FilterMap(items, func(item backend.Item) (types.Instance, bool) {
+		var instance types.InstanceV1
+		if err := utils.FastUnmarshal(item.Value, &instance); err != nil {
+			s.log.Warnf("Skipping instance at %s, failed to unmarshal: %v", item.Key, err)
+			return nil, false
+		}
+		if err := instance.CheckAndSetDefaults(); err != nil {
+			s.log.Warnf("Skipping instance at %s: %v", item.Key, err)
+			return nil, false
+		}
+		if !req.Match(&instance) {
+			return nil, false
+		}
+		return &instance, true
+	})
+}
+
+// GetRawInstance gets an instance resource by server ID.
+func (s *PresenceService) GetRawInstance(ctx context.Context, serverID string) (types.Instance, []byte, error) {
+	item, err := s.Get(ctx, backend.Key(instancePrefix, serverID))
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, nil, trace.NotFound("failed to locate instance %q", serverID)
+		}
+		return nil, nil, trace.Wrap(err)
+	}
+
+	var instance types.InstanceV1
+	if err := utils.FastUnmarshal(item.Value, &instance); err != nil {
+		return nil, nil, trace.BadParameter("failed to unmarshal instance %q: %v", serverID, err)
+	}
+
+	if err := instance.CheckAndSetDefaults(); err != nil {
+		return nil, nil, trace.BadParameter("instance %q appears malformed: %v", serverID, err)
+	}
+
+	return &instance, item.Value, nil
+}
+
+// CompareAndSwapInstance creates or updates the underlying instance resource based on the currently
+// expected value. The first call to this method should use the value returned by GetRawInstance for the
+// 'expect' parameter. Subsequent calls should use the value returned by this method.
+func (s *PresenceService) CompareAndSwapInstance(ctx context.Context, instance types.Instance, expect []byte) ([]byte, error) {
```

Comment:
Should we use this pattern for other CASable resources as well? `CompareAndSwapCertAuthority` comes to mind.

Reply:
All else being equal, I do prefer this pattern.  Not sure it's worth the trouble for Cert Authorities though.  I only did this because eliminating the extra read has a big effect at the scale of instance resources.  Wouldn't make a big difference for CAs.

Reply:
In theory this pattern is more correct.  I'd be hesitant to make it the default as things stand.  The perf benefits only matter if you're dealing w/ a resource that scales w/ cluster size, and it might work inconsistently or not at all when dealing w/ cached items.